changes/2023_7_12_update-keystore-structure/background.md

```diff
@@ -0,0 +1,286 @@
+
+### Store the custom values in a single attribute
+
+This makes writing a policy based on these values overly complicated.
+Since there is no way to have granular control of the values.
+A customer may want to require several custom key/values.
+But different roles have difference accesses.
+KMS policy control is only on strict equality so combining all this information make it easy for our development, but more complicated for customers to use.
+
+## How is the encryption context for access propagated to versions?
+
+### _Authenticate the values from the current active key_
+
+We do not want to give the versioning process access to the key, so decrypt is out.
+But we already need reEncrypt to create the active entry.
+We could reEncrypt the existing entry to authenticate the existing encryption context.
+This means performing an additional correctness step, but this process is not in the hot path and does not occur frequently.
+If this reEncrypt does not change the encryption context of the new blob, then the scope of policy protecting the branch keys can not be escaped.
+This means that additional policy control could be written to constrain this process so that it can only reEncrypt from a version item to an active item and from active item to the same active item.
+But this policy is not required to function, it is just an exercise in least privilege.
+
+The idea here would be
+
+1. Get the current item (if no active item exists fail)
+2. ReEncrypt this item to encryption context identical to the existing encryption context.
+3. All encryption context is now authenticated
+4. Construct a `GenerateDataKeyWithoutPlaintext` request with the now authenticated additional values
+5. ReEncrypt this new version so that a new Active item can be inserted
+6. Insert both new items as a transaction
+
+### Get the values from an unauthenticated meta entry
+
+No, having this data be unauthenticated, but then add it to an authenticated process is crazy.
+
+### Just use the existing values on the active entry without authentication
+
+No, this is less dangerous than just using unauthenticated data, but it creates an authentication race. Having a way to inject unauthenticated data into an authenticated process is a bad idea.
+
+### Get the values from an authenticated meta entry
+
+The authentication could be done with decrypt, but this would give decrypt to the create/rotation process.
+This could be constrained with policy, but this is not correct by construction.
+If a customer does not apply this policy then the process is more privileged.
+If it uses the same reEncrypt as above, then what additional value does this provide?
+It also creates a new value that is not a branch key that is now floating around.
+
+## How can we let customers control the branch key id?
+
+### _Require an encryption context constraint when creating with a defined branch key id_
```

Comment:
```suggestion
### _Require that encryption context MUST also be provided when creating with a defined branch key id_
```

Something like this; as far as I understand it's a "constraint that encryption context must be provided" and I think this makes it clearer...